[Ruby] Add enumeration Concept Exercise (#1928)
exercism/v3: languages/ruby/exercises/concept/enumeration/.docs/introduction.md

Enumeration is the act of stepping through a collection (Array, Hash, etc) and performing some action on each object.

> Enumeration is a key concept in Ruby and is used for sorting (`sort_by`), grouping (`group_by`), mapping (`map`), reducing (`reduce`), and much more. You'll most frequently see enumeration as the idiomatic way iterating through collections rather than using loops.

Author: Happy to remove these if we leave the doc in below. Originally I had these and not the docs link, but now I have the docs link, I'm happy to remove one or the other. 

I think we need to either specify hints about the methods, or links to them. In the Elxiir, one for example, the methods are much more explicitely stated as part of the learning goal, whereas here the methods are more incidental to the learning goal of how enums + blocks work. So happy to remove this if I'm allowed the link below :)

> Enumeration is a key concept in Ruby and is used for sorting (`sort_by`), grouping (`group_by`), mapping (`map`), reducing (`reduce`), and much more. You'll most frequently see enumeration as the idiomatic way iterating through collections rather than using loops.

Reviewer: If we keep a short list, perhaps we don't need the "white lie" about `each` as we have examples of `Enumerable` methods that we could chain instead? Even if we link it, it is a short step from providing one of them for the chaining example to the link we encourage them to read.

> Enumeration is a key concept in Ruby and is used for sorting (`sort_by`), grouping (`group_by`), mapping (`map`), reducing (`reduce`), and much more. You'll most frequently see enumeration as the idiomatic way iterating through collections rather than using loops.

Reviewer 2: > I think we need to either specify hints about the methods, or links to them. In the Elxiir, one for example, the methods are much more explicitely stated as part of the learning goal, whereas here the methods are more incidental to the learning goal of how enums + blocks work. So happy to remove this if I'm allowed the link below :)

My preference would be for these specific mentions over the link. See my other comment for my reasoning. I think part of the issue here is that the exercise is less "guided," in that there is less of a clear path to solve the issue than in most other exercises. This comes with the subject, so _some_ guidance can definitely help.

> Enumeration is a key concept in Ruby and is used for sorting (`sort_by`), grouping (`group_by`), mapping (`map`), reducing (`reduce`), and much more. You'll most frequently see enumeration as the idiomatic way iterating through collections rather than using loops.

Reviewer 2: So Jeremy and I have discussed this and we've decided that we want to keep both the link to the `Enumerable` module's documentation _and_ the listing of the individual methods (`group_by` etc.) The reasons for that are:

1. This exercise is relatively exceptional in that it teaches a bunch of methods at once and students can opt to implement their solution using completely different methods. This makes the possible implementations huge, so restricting that makes sense.
1. The official Ruby docs are not the first thing one finds when googling for the methods, which can lead to students using documents we don't want them to use.

All in all this is a relatively unique case where it does make sense to have the link to the documentation and the listing of the methods.

A simple enumeration to print each word in an array would look like this:

```ruby
words = %w[the cat sat on the mat]
words.each { |word| print word }

#=> the
#=> cat
#=> ...
#=> mat
```

> In this example, we have called the Array's `each` method and passed in a _block_, which takes one parameter (`word`) and uppercases it. Blocks are small groupings of statements that can be executed multiple times. They can be thought of as closures or anonymous functions. Blocks are defined using the `do...end` syntax (above), or the `{}` (below). The styles are interchanagable and differening opinions exist about when each should be used.

Author: A linter changed it 🤦

> We can also chain enumarable methods. For example, we can chain `.with_index` onto `each` to print out the index of an object as well as it's value:

Author: I'm going to leave this one as a "white lie" I think, as I don't want to give away the other methods for the exercise.

> We can also chain enumarable methods. For example, we can chain `.with_index` onto `each` to print out the index of an object as well as it's value:

Reviewer: We could use map on map.  As we already reveal that.  If map is used already, it qualifies.  And I think we reveal that already.

```ruby
words = %w[the cat sat on the mat]
words.each.with_index { |word, index| print "#{index}: #{word}" }

#=> 0. the
#=> 1. cat
#=> ...
#=> 5. mat
```

Enumerating hashes is exactly the same as enumerating arrays, except that the block receives two parameters: the key and the value:

> ```

Reviewer 2: ```suggestion
```ruby
```

pet_names = {cat: "bob", horse: "caris", mouse: "arya"}
words.each { |animal, name| ... }

# The two params should be put in brakets when chaining
words.each.with_index { |(animal, name), index| ... }
```

> The various enumeration methods can be see in the docs for the the [Enumerable Module](https://ruby-doc.org/core-2.7.1/Enumerable.html) which is part of Arrays, Hashes, etc.

Reviewer: ```suggestion
The various enumeration methods can be seen in the docs for the the [Enumerable Module](https://ruby-doc.org/core-2.7.1/Enumerable.html) which is part of Arrays, Hashes, etc.
```

> The various enumeration methods can be see in the docs for the the [Enumerable Module](https://ruby-doc.org/core-2.7.1/Enumerable.html) which is part of Arrays, Hashes, etc.

Author: I think the intention with that quote is that we're not pushing people to have to read other blogs to understand the concept. I think linking to the official docs, which contain the methods that someone needs to discover to solve the exercise, is a "friendly" action rather than "harmful" one. I **believe** I've been consistent in that opinion, but happy to be told where I'm being a total hypocrite :)

> The various enumeration methods can be see in the docs for the the [Enumerable Module](https://ruby-doc.org/core-2.7.1/Enumerable.html) which is part of Arrays, Hashes, etc.

Reviewer 2: > I think linking to the official docs, which contain the methods that someone needs to discover to solve the exercise, is a "friendly" action rather than "harmful" one.

Well, in the case of Ruby documentation, I feel like I could make the case that it is harmful :P My main "worry" here is that we link to the entire document, not specific methods (which we _can_ do in the `hints.md` file). There is just so much content there that I feel it could easily feel overwhelming.

Maybe I remember this wrong, but I thought we wanted to mention the official names but not link to them here, but only in the `hints.md` file. This is also how I've reviewed all the existing tracks' PR's, so I we change this we should probably be making this more clear in the documentation.

> The various enumeration methods can be see in the docs for the the [Enumerable Module](https://ruby-doc.org/core-2.7.1/Enumerable.html) which is part of Arrays, Hashes, etc.

Reviewer 2: But maybe I'm overthinking this and linking to documentation is fine.

Enumeration is the act of stepping through a collection (Array, Hash, etc) and performing some action on each object.
Enumeration is a key concept in Ruby and is used for sorting (`sort_by`), grouping (`group_by`), mapping (`map`), reducing (`reduce`), and much more. You'll most frequently see enumeration as the idiomatic way iterating through collections rather than using loops.

A simple enumeration to print each word in an array would look like this:

```ruby
words = %w[the cat sat on the mat]
words.each do |word| 
  print word 
end

#=> the
#=> cat
#=> ...
#=> mat

> ```

Reviewer 2: I was slightly confused about the `#=> ...`, maybe use less words or list all the outputs?

> In this example, we have called the Array's `each` method and passed in a _block_, which takes one parameter (`word`) and prints it. Blocks are small groupings of statements that can be executed multiple times. They can be thought of as closures or anonymous functions. Blocks are defined using the `do...end` syntax (above), or the `{}` (below). The styles are interchangeable and differing opinions exist about when each should be used.

Reviewer: ```suggestion
In this example, we have called the Array's `each` method and passed in a _block_, which takes one parameter (`word`) and prints it. Blocks are small groupings of statements that can be executed multiple times. They can be thought of as closures or anonymous functions. Blocks are defined using the `do...end` syntax (above), or the `{}` (below). The styles are interchangeable and differing opinions exist about when `each` should be used.
```

We can also chain enumerable methods. For example, we can chain `.with_index` onto `each` to print out the index of an object as well as it's value:

```ruby
words = %w[the cat sat on the mat]

> words.each.with_index { |word, index| print "#{index}: #{word}" }

Reviewer: ```suggestion
words.map.with_index { |word, index| puts "#{index}: #{word}" }
```
Which will output:

```
0: the
1: cat
2: sat
3: on
4: the
5: mat
```
The colon is used here to avoid potential recognition of the numbers as "with a period".

But now "chain enumerable methods" is more accurate in what is happening here, we chain `map` (an `Enumerable` method) to `with_index` (another `Enumerable` method).

#=> 0. the
#=> 1. cat
#=> ...
#=> 5. mat
```

> Enumerating hashes is exactly the same as enumerating arrays, except that the block receives two parameters: the key and the value:

Reviewer: ```suggestion
Enumerating `Hash` objects is exactly the same as enumerating `Array` objects, except that the block receives two parameters: the key and the value:
```

In Ruby (and computer science) the word "hash" is used.  Here we are specifically speaking of `Hash` objects, as opposed to the `hash` that is required to be callable on anything that can be a key for a `Hash`.  We don't want to mention that here, but we also don't want to confuse the terms, so being explicit that we are iterating a `Hash` is better than speaking loosely about it.  And since the documentation is linked, the documentation also speaks about "hash" (it is a method as well), and it talks about how or why the hash is used.

```ruby
pet_names = {cat: "bob", horse: "caris", mouse: "arya"}
words.each { |animal, name| ... }

> # The two params should be put in brakets when chaining

Reviewer: I prefer "arguments" to `params`.  Surely, we should prefer `parameters` if we _must_ use that word in terms of methods, as opposed to "limits to a function."

words.each.with_index { |(animal, name), index| ... }
```

> The various enumeration methods can be seen in the docs for the the [Enumerable Module](https://ruby-doc.org/core-2.7.1/Enumerable.html) which is part of Arrays, Hashes, etc.

Reviewer: ```suggestion
The `Enumerable` methods can be seen in the docs for the the [Enumerable Module](https://ruby-doc.org/core-2.7.1/Enumerable.html).  Array, Hash, and any object that needs to be enumerated can include this module to gain those methods.
```

Reviewer: I add this because `Enumerator` also has some enumeration methods.  We won't necessarily see those methods in the `Enumerable` documentation, but we will see the `Enumerable` methods there.
